source/utils/helpers.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"fmt"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+
+	"sigs.k8s.io/external-dns/endpoint"
+)
+
+func EndpointsForHostname(hostname string, targets endpoint.Targets, ttl endpoint.TTL, providerSpecific endpoint.ProviderSpecific, setIdentifier string, resource string) []*endpoint.Endpoint {
+	var endpoints []*endpoint.Endpoint
+
+	var aTargets endpoint.Targets
+	var aaaaTargets endpoint.Targets
+	var cnameTargets endpoint.Targets
+
+	for _, t := range targets {
+		switch suitableType(t) {
+		case endpoint.RecordTypeA:
+			aTargets = append(aTargets, t)
+		case endpoint.RecordTypeAAAA:
+			aaaaTargets = append(aaaaTargets, t)
+		default:
+			cnameTargets = append(cnameTargets, t)
+		}
+	}
+
+	if len(aTargets) > 0 {
+		epA := endpoint.NewEndpointWithTTL(hostname, endpoint.RecordTypeA, ttl, aTargets...)
+		if epA != nil {
+			epA.ProviderSpecific = providerSpecific
+			epA.SetIdentifier = setIdentifier
+			if resource != "" {
+				epA.Labels[endpoint.ResourceLabelKey] = resource
+			}
+			endpoints = append(endpoints, epA)
+		}
+	}
+
+	if len(aaaaTargets) > 0 {
+		epAAAA := endpoint.NewEndpointWithTTL(hostname, endpoint.RecordTypeAAAA, ttl, aaaaTargets...)
+		if epAAAA != nil {
+			epAAAA.ProviderSpecific = providerSpecific
+			epAAAA.SetIdentifier = setIdentifier
+			if resource != "" {
+				epAAAA.Labels[endpoint.ResourceLabelKey] = resource
+			}
+			endpoints = append(endpoints, epAAAA)
+		}
+	}
+
+	if len(cnameTargets) > 0 {
+		epCNAME := endpoint.NewEndpointWithTTL(hostname, endpoint.RecordTypeCNAME, ttl, cnameTargets...)
+		if epCNAME != nil {
+			epCNAME.ProviderSpecific = providerSpecific
+			epCNAME.SetIdentifier = setIdentifier
+			if resource != "" {
+				epCNAME.Labels[endpoint.ResourceLabelKey] = resource
+			}
+			endpoints = append(endpoints, epCNAME)
+		}
+	}
+
+	return endpoints
+}
+
+func ProviderSpecificAnnotations(annotations map[string]string) (endpoint.ProviderSpecific, string) {
```

Comment:
Instead of set those annotations in constants.go of utils package (which is confusing and unexpected), wdyt about passing an array of those specific annotations in arg of this func ?

Reply:
Or just taking the prefix, like for the other ones ?

Reply:
We have a proposal waiting https://github.com/kubernetes-sigs/external-dns/pull/5080. Annotations are just everywhere, and is need revisit. What you suggesting make sense, is just going to be logic rewrite. 

I could create for now folder `source/annotations/annotations.go` as a temporary solution?

Reply:
Could move all annotations related logic to this `source/annottions/...` folder. Will this work?

Reply:
Could move all annotations related logic that is in `source/utils` to this source/annottions/... folder. Will this work?

Reply:
Have a look. At the moment, due to not having separate folders, there is no isolation in between code. If we move all sources into it's folder, this would make it possible to import things correctly without cyclic dependencies.

Reply:
Have a look, made a change. At the moment, due to not having separate folders, there is no isolation in between code. If we move all sources into it's folder, this would make it possible to import things correctly without cyclic dependencies.

Reply:
Have a look, made a change. At the moment, due to not having separate folders, there is no isolation in between code. If we move all sources into it's folder, same as providers, this would make it possible to import things correctly without cyclic dependencies.